src/basic/terminal-util.c

```diff
@@ -891,33 +891,24 @@ bool on_tty(void) {
 }
 
 int getttyname_malloc(int fd, char **ret) {
-        size_t l = 100;
+        char path[PATH_MAX], *c; /* PATH_MAX is counted *with* the trailing NUL byte */
         int r;
 
         assert(fd >= 0);
         assert(ret);
 
-        for (;;) {
-                char path[l];
-
-                r = ttyname_r(fd, path, sizeof(path));
-                if (r == 0) {
-                        char *c;
+        r = ttyname_r(fd, path, sizeof path); /* positive error */
+        assert(r >= 0);
+        if (r == ERANGE)
+                return -ENAMETOOLONG;
```

Comment:
hmm, this is an interesting tweak. It makes me wonder: should we do that for readlink_malloc() too? (after all ttyname_r() is just a glorified wrapper around readlink) i.e. would it make sense to always always ensure that paths we retrieve from the kernel are not longer than PATH_MAX? or does the kernel already enforce the limit anyway, and this is redundant?

(At least for getwd() it's documented that the current working directory could be longer than PATH_MAX...)

Reply:
If I'm looking in the right place, `getname_flags()` which is called by `getname()` which is called by `do_symlinkat()` which implements `symlink()` and `symlinkat()`, enforces a hard limit of `PATH_MAX-1`. So it would seem it's not possible to *create* a symlink with a very long name. But `readlink()` calls `vfs_readlink()` which offloads to filesystem-specific implementations, so its hard to say for sure if they set the limits. In particular, `readlink()` can be used to look up AFS mountpoints.